Tablet config for ELAN 2627

daenney opened the pull request:
This model is found in the HP Envy x360 ag0xxxx series.

I have a few questions though:
* Width and height, couldn't really find them specified by HP so I just used a ruler and rounded up to the closest amount of inches. Does that make sense or should I round down to 11 and 6 instead?
* The device class is giving me trouble. `PenDisplay` seems the most appropriate but if I use that the Pen doesn't work at all. Switching to Intuos4 makes the pen work but the tracking is bonkers slow, the pointer lags behind the pen a great deal, almost a second and its location is off. I also tried `Class=ISDV4` which seems to get me the same result as `Class=Intuos4`. Which one should I pick?
* Touching the screen (with my finger) seems to result in multiple events. So for example when I touch the header bar of an app it goes full-screen b/c Gnome apparently believes I double-clicked it. I'm not sure if this has anything to do with libwacom or if I should raise this with the libinput folks

When I run `libwacom-list-local-devices` I get 2, contrary to what I would expect:
```
[Device]
Name=ELAN 2627
DeviceMatch=i2c:04f3:2627;
Class=PenDisplay
Width=12
Height=7
IntegratedIn=Display;System;
Styli=0xfffff;0xffffe;

[Features]
Reversible=false
Stylus=true
Ring=false
Ring2=false
Touch=true
TouchSwitch=false
StatusLEDs=
NumStrips=0
Buttons=0
---------------------------------------------------------------
[Device]
Name=Wacom Serial Tablet WACf004
DeviceMatch=serial:0000:0000;
Class=ISDV4
Width=0
Height=0
IntegratedIn=Display;System;
Styli=0xfffff;0xffffe;

[Features]
Reversible=false
Stylus=true
Ring=false
Ring2=false
Touch=false
TouchSwitch=false
StatusLEDs=
NumStrips=0
Buttons=0
---------------------------------------------------------------
```
Is the `Wacom Serial Tablet WACf004` device supposed to be there?

Despite adding this config devices didn't show up properly for X11, I had to add this snippet:

```
Section "InputClass"
    Identifier "Elan driver override"
    MatchUSBID "04f3:*"
    MatchDevicePath "/dev/input/event*"
    MatchIsTablet "true"
    Driver "wacom"
EndSection
```

I get this from `xinput`:
```
⎡ Virtual core pointer                          id=2    [master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer                id=4    [slave  pointer  (2)]
⎜   ↳ ELAN0732:00 04F3:2627 stylus              id=11   [slave  pointer  (2)]
⎜   ↳ ELAN0732:00 04F3:2627                     id=12   [slave  pointer  (2)]
⎜   ↳ SynPS/2 Synaptics TouchPad                id=14   [slave  pointer  (2)]
⎜   ↳ ELAN0732:00 04F3:2627 eraser              id=17   [slave  pointer  (2)]
⎣ Virtual core keyboard                         id=3    [master keyboard (2)]
    ↳ Virtual core XTEST keyboard               id=5    [slave  keyboard (3)]
    ↳ Power Button                              id=6    [slave  keyboard (3)]
    ↳ Video Bus                                 id=7    [slave  keyboard (3)]
    ↳ Power Button                              id=8    [slave  keyboard (3)]
    ↳ HP Wide Vision HD Camera: HP Wi           id=9    [slave  keyboard (3)]
    ↳ HP IR Camera: HP IR Camera                id=10   [slave  keyboard (3)]
    ↳ AT Translated Set 2 keyboard              id=13   [slave  keyboard (3)]
    ↳ HP WMI hotkeys                            id=15   [slave  keyboard (3)]
    ↳ HP Wireless hotkeys                       id=16   [slave  keyboard (3)]
```

And `xsetwacom --list devices`:
```
ELAN0732:00 04F3:2627 stylus            id: 11  type: STYLUS    
ELAN0732:00 04F3:2627 eraser            id: 17  type: ERASER  
```

I took a look at the different properties for the devices xinput lists.

### Class=PenDisplay
* ELAN0732:00 04F3:2627 (id=12)
```
Device 'ELAN0732:00 04F3:2627':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix (311):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix Default (312):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Send Events Modes Available (267):     1, 0
        libinput Send Events Mode Enabled (268):        0, 0
        libinput Send Events Mode Enabled Default (269):        0, 0
        Device Node (270):      "/dev/input/event6"
        Device Product ID (271):        1267, 9767
```
* ELAN0732:00 04F3:2627 stylus (id=11)
```
Device 'ELAN0732:00 04F3:2627 stylus':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        Device Accel Profile (273):     0
        Device Accel Constant Deceleration (274):       1.000000
        Device Accel Adaptive Deceleration (275):       1.000000
        Device Accel Velocity Scaling (276):    10.000000
        Device Node (270):      "/dev/input/event7"
        Wacom Tablet Area (283):        0, 0, 18460, 10400
        Wacom Rotation (284):   0
        Wacom Pressurecurve (285):      0, 0, 100, 100
        Wacom Serial IDs (286): 9767, 0, 2, 0, 0
        Wacom Serial ID binding (287):  0
        Wacom Pressure Threshold (288): 26
        Wacom Sample and Suppress (289):        2, 4
        Wacom Enable Touch (290):       0
        Wacom Hover Click (291):        1
        Wacom Enable Touch Gesture (292):       0
        Wacom Touch Gesture Parameters (293):   0, 0, 250
        Wacom Tool Type (294):  "STYLUS" (272)
        Wacom Button Actions (295):     "Wacom button action 0" (296), "Wacom button action 1" (297), "Wacom button action 2" (298), "None" (0), "None" (0), "None" (0), "None" (0), "Wacom button action 3" (299)
        Wacom button action 0 (296):    1572865
        Wacom button action 1 (297):    1572866
        Wacom button action 2 (298):    1572867
        Wacom button action 3 (299):    1572872
        Wacom Pressure Recalibration (300):     1
        Wacom Panscroll Threshold (301):        819
        Device Product ID (271):        1267, 9767
        Wacom Debug Levels (302):       0, 0
```

* ELAN0732:00 04F3:2627 eraser (id=17)
```
Device 'ELAN0732:00 04F3:2627 eraser':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        Device Accel Profile (273):     0
        Device Accel Constant Deceleration (274):       1.000000
        Device Accel Adaptive Deceleration (275):       1.000000
        Device Accel Velocity Scaling (276):    10.000000
        Device Node (270):      "/dev/input/event7"
        Wacom Tablet Area (283):        0, 0, 18460, 10400
        Wacom Rotation (284):   0
        Wacom Pressurecurve (285):      0, 0, 100, 100
        Wacom Serial IDs (286): 9767, 0, 10, 0, 0
        Wacom Serial ID binding (287):  0
        Wacom Pressure Threshold (288): 26
        Wacom Sample and Suppress (289):        2, 4
        Wacom Enable Touch (290):       0
        Wacom Enable Touch Gesture (292):       0
        Wacom Touch Gesture Parameters (293):   0, 0, 250
        Wacom Tool Type (294):  "ERASER" (340)
        Wacom Button Actions (295):     "Wacom button action 0" (296), "Wacom button action 1" (297), "Wacom button action 2" (298), "None" (0), "None" (0), "None" (0), "None" (0), "Wacom button action 3" (299)
        Wacom button action 0 (296):    1572865
        Wacom button action 1 (297):    1572866
        Wacom button action 2 (298):    1572867
        Wacom button action 3 (299):    1572872
        Wacom Pressure Recalibration (300):     1
        Wacom Panscroll Threshold (301):        819
        Device Product ID (271):        1267, 9767
        Wacom Debug Levels (302):       0, 0
```

Stylus and eraser appear to be the same, which I suppose is expected. I'm a bit puzzled by the other device, `id=12`.

### Class=Intuos4
* ELAN0732:00 04F3:2627 (id=12)
```
Device 'ELAN0732:00 04F3:2627':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix (311):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix Default (312):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Send Events Modes Available (267):     1, 0
        libinput Send Events Mode Enabled (268):        0, 0
        libinput Send Events Mode Enabled Default (269):        0, 0
        Device Node (270):      "/dev/input/event6"
        Device Product ID (271):        1267, 9767
```
* ELAN0732:00 04F3:2627 stylus (id=11)
```
Device 'ELAN0732:00 04F3:2627':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix (311):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Calibration Matrix Default (312):      1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Send Events Modes Available (267):     1, 0
        libinput Send Events Mode Enabled (268):        0, 0
        libinput Send Events Mode Enabled Default (269):        0, 0
        Device Node (270):      "/dev/input/event6"
        Device Product ID (271):        1267, 9767
daenney@s-monocerotis ~> cat xinput-stylus-2.txt 
Device 'ELAN0732:00 04F3:2627 stylus':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        Device Accel Profile (273):     0
        Device Accel Constant Deceleration (274):       1.000000
        Device Accel Adaptive Deceleration (275):       1.000000
        Device Accel Velocity Scaling (276):    10.000000
        Device Node (270):      "/dev/input/event7"
        Wacom Tablet Area (283):        0, 0, 18460, 10400
        Wacom Rotation (284):   0
        Wacom Pressurecurve (285):      0, 0, 100, 100
        Wacom Serial IDs (286): 9767, 1, 2, 1, 2
        Wacom Serial ID binding (287):  0
        Wacom Pressure Threshold (288): 26
        Wacom Sample and Suppress (289):        2, 4
        Wacom Enable Touch (290):       0
        Wacom Hover Click (291):        1
        Wacom Enable Touch Gesture (292):       0
        Wacom Touch Gesture Parameters (293):   0, 0, 250
        Wacom Tool Type (294):  "STYLUS" (272)
        Wacom Button Actions (295):     "Wacom button action 0" (296), "Wacom button action 1" (297), "Wacom button action 2" (298), "None" (0), "None" (0), "None" (0), "None" (0), "Wacom button action 3" (299)
        Wacom button action 0 (296):    1572865
        Wacom button action 1 (297):    1572866
        Wacom button action 2 (298):    1572867
        Wacom button action 3 (299):    1572872
        Wacom Pressure Recalibration (300):     1
        Wacom Panscroll Threshold (301):        819
        Device Product ID (271):        1267, 9767
        Wacom Debug Levels (302):       0, 0
```

* ELAN0732:00 04F3:2627 eraser (id=17)
```
Device 'ELAN0732:00 04F3:2627 eraser':
        Device Enabled (145):   1
        Coordinate Transformation Matrix (147): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        Device Accel Profile (273):     0
        Device Accel Constant Deceleration (274):       1.000000
        Device Accel Adaptive Deceleration (275):       1.000000
        Device Accel Velocity Scaling (276):    10.000000
        Device Node (270):      "/dev/input/event7"
        Wacom Tablet Area (283):        0, 0, 18460, 10400
        Wacom Rotation (284):   0
        Wacom Pressurecurve (285):      0, 0, 100, 100
        Wacom Serial IDs (286): 9767, 0, 10, 0, 0
        Wacom Serial ID binding (287):  0
        Wacom Pressure Threshold (288): 26
        Wacom Sample and Suppress (289):        2, 4
        Wacom Enable Touch (290):       0
        Wacom Enable Touch Gesture (292):       0
        Wacom Touch Gesture Parameters (293):   0, 0, 250
        Wacom Tool Type (294):  "ERASER" (340)
        Wacom Button Actions (295):     "Wacom button action 0" (296), "Wacom button action 1" (297), "Wacom button action 2" (298), "None" (0), "None" (0), "None" (0), "None" (0), "Wacom button action 3" (299)
        Wacom button action 0 (296):    1572865
        Wacom button action 1 (297):    1572866
        Wacom button action 2 (298):    1572867
        Wacom button action 3 (299):    1572872
        Wacom Pressure Recalibration (300):     1
        Wacom Panscroll Threshold (301):        819
        Device Product ID (271):        1267, 9767
        Wacom Debug Levels (302):       0, 0
```

The only difference appears to be a change in Wacom Serial IDs on the stylus:
```diff
12c12
<       Wacom Serial IDs (286): 9767, 0, 2, 0, 0
---
>       Wacom Serial IDs (286): 9767, 1, 2, 1, 2
```

I get a similar diff on the stylus between PenDisplay and ISDV4:
```diff
12c12
<       Wacom Serial IDs (286): 9767, 0, 2, 0, 0
---
>       Wacom Serial IDs (286): 9767, 1, 0, 1, 0
```

daenney commented:
CI fails on `test-tablet-validity` so I assume it's something I did. Unfortunately it doesn't tell me what is wrong.

daenney commented:
One other thing of note; I cannot get this device to work on any Linux version prior to 5.2 (currently at rc3). I assume this is due to a change to i2c but I've yet to be able to track down the commit in question.

daenney commented:
Please ignore all of the above. Switching the config back to `Class=PenDisplay` with `Styli=0xffffd` resolves both the slow tracking and the touch input being detected multiple times problems.

This makes the device fully functional and usable in things like Krita too.

whot commented:
>  so I just used a ruler...

yep, that's good enough. This is more for an approximation/human printing than anything else anyway.

>  Switching to Intuos4 makes the pen work but the tracking is bonkers slow,

That is ... strange. libwacom has no effect on how the device itself works, so this shouldn't do anything. Doubly so because afaik nothing uses the `Class` specifier anyway. And triply so because when you use the wacom xorg driver - it doesn't actually use libwacom and doesn't care about any of those configurations. So this is really strange and I wonder if there's some other bug that triggers. Did you check the journal? maybe something is getting confused by the device and using cpu time? or maybe the device is susceptible to reboots/re-inits and you just happened to hit it while changing libwacom files?

FWIW, traditionally we used ISDV4 for those devices, see e.g. `data/elan-264c.tablet` which is for another HP Envy x360.

> The only difference appears to be a change in Wacom Serial IDs on the stylus:

that only shows the most recent stylus and changes whenever a pen is in proximity. any changes here are usually safe to ignore.

daenney commented:
It looks like there's some other bug that triggers. Applying this same configuration on a clean installation of the same device, same kernel version, xorg, libwacom etc. and I'm now back to dog slow/out of sync tracking.

I'm wondering how to figure out what's up with this. Clearly it's possible for it to work to perfection as it accidentally happened before.

Since it seems to have no effect, shall I switch this back to ISDV4?

daenney commented:
One question I still have; is it expected that `libwacom-list-local-devices` shows 2 devices? It seems really odd to me, like it's not detecting something properly.

daenney commented:
Alright, I managed to find the bug by pure chance. It appears to be the result of a badly initialised ACPI register. Forcefully powerding down the device and then holding the power button a couple of seconds seems to undo whatever it is and then booting it everything works. I accidentally performed this sequence before when testing b/c at some point I did something that crashed the touchpad. This appears to be some kind of interplay between Windows and Linux as I had the device originally booted in Windows in order to perform the necessary firmware upgrades.

I've updated my PR and switched it back to ISDV4, everything works as expected now.

jigpu commented:
> One question I still have; is it expected that `libwacom-list-local-devices` shows 2 devices? It seems really odd to me, like it's not detecting something properly.

Just to follow-up on this particular question... The second device shows up because there's a (probably purely internal) serial port device somewhere in the system. There's no standard way to ask serial devices exactly what they are, but since some table sensors connect via serial, libwacom has to err on the side of caution and say that it found something (even though it almost certainly has not).

whot commented:
@daenney can you try the patch #96 please, that should give you an indication where this comes from. I looked at the `libwacom-list-local-devices` source and it only looks at event nodes. Which would indicate that xf86-input-wacom's `wacom.rules` file triggers the inputattach service and creates an event node.

A `udevadm info --export-db` output should tell us more here.

daenney commented:
The output of `--export-db` is massive, 6k lines. Any particular thing you're looking for that I can search for?

daenney commented:
Applying #96 I get:

```
-------------------------------------------------- /dev/input/event7
[Device]
Name=ELAN 2627
ModelName=
DeviceMatch=i2c:04f3:2627;
Class=ISDV4
Width=12
Height=7
IntegratedIn=Display;System;
Styli=0xfffff;0xffffe;

[Features]
Reversible=false
Stylus=true
Ring=false
Ring2=false
Touch=true
TouchSwitch=false
StatusLEDs=
NumStrips=0
Buttons=0
-------------------------------------------------- /dev/input/event13
[Device]
Name=Wacom Serial Tablet WACf004
ModelName=
DeviceMatch=serial:0000:0000;
Class=ISDV4
Width=0
Height=0
IntegratedIn=Display;System;
Styli=0xfffff;0xffffe;

[Features]
Reversible=false
Stylus=true
Ring=false
Ring2=false
Touch=false
TouchSwitch=false
StatusLEDs=
NumStrips=0
Buttons=0
```

```
   ↳ AT Translated Set 2 keyboard              id=13   [slave  keyboard (3)]
```

whot commented:
Note that the X ids have nothing to do with the event nodes underneath (one event node may have multiple X devices, and the sort order differs too). You'd have to look at the `Device Node` xinput property (`xinput list-props <device name>`) to make sure you're looking at the right device.

And yes, the `udevadm` output is large but it's a lot easier for us to look through it than describing what exactly to look for (because I know what to look for but I don't know how to explain it :)

daenney commented:
Right. Looks like it's picking up on my touchpad:

```
-------------------------------------------------- /dev/input/event10
[Device]
Name=Wacom Serial Tablet WACf004
ModelName=
DeviceMatch=serial:0000:0000;
Class=ISDV4
Width=0
Height=0
IntegratedIn=Display;System;
Styli=0xfffff;0xffffe;

[Features]
Reversible=false
Stylus=true
Ring=false
Ring2=false
Touch=false
TouchSwitch=false
StatusLEDs=
NumStrips=0
Buttons=0
```

```
Device 'SynPS/2 Synaptics TouchPad':
        Device Enabled (146):   1
        Coordinate Transformation Matrix (148): 1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
        libinput Tapping Enabled (314): 1
        libinput Tapping Enabled Default (315): 0
        libinput Tapping Drag Enabled (316):    1
        libinput Tapping Drag Enabled Default (317):    1
        libinput Tapping Drag Lock Enabled (318):       0
        libinput Tapping Drag Lock Enabled Default (319):       0
        libinput Tapping Button Mapping Enabled (320):  1, 0
        libinput Tapping Button Mapping Default (321):  1, 0
        libinput Natural Scrolling Enabled (322):       1
        libinput Natural Scrolling Enabled Default (323):       0
        libinput Disable While Typing Enabled (324):    1
        libinput Disable While Typing Enabled Default (325):    1
        libinput Scroll Methods Available (326):        1, 1, 0
        libinput Scroll Method Enabled (327):   1, 0, 0
        libinput Scroll Method Enabled Default (328):   1, 0, 0
        libinput Click Methods Available (329): 1, 1
        libinput Click Method Enabled (330):    0, 1
        libinput Click Method Enabled Default (331):    1, 0
        libinput Middle Emulation Enabled (332):        0
        libinput Middle Emulation Enabled Default (333):        0
        libinput Accel Speed (334):     0.400000
        libinput Accel Speed Default (335):     0.000000
        libinput Left Handed Enabled (336):     0
        libinput Left Handed Enabled Default (337):     0
        libinput Send Events Modes Available (268):     1, 1
        libinput Send Events Mode Enabled (269):        0, 0
        libinput Send Events Mode Enabled Default (270):        0, 0
        Device Node (271):      "/dev/input/event10"
        Device Product ID (272):        2, 7
        libinput Drag Lock Buttons (338):       <no items>
        libinput Horizontal Scroll Enabled (339):       1
```

daenney commented:
Here's the udev db: https://gist.github.com/daenney/6612a960eaa2887575fe6a8f778cbd87
